### DISCUSSION TITLE
Queues with different OS versions

### DISCUSSION THREAD

**Discussion by marianne013** [DIRACGrid/DIRAC#7321] [Q&A] [answered]
Hi,

RAL-LCG2 now has queues with different operating systems: EL7,8 & 9 (you can see them appearing on the certification server). 
Is there any way I can use the platform/os matching in DIRAC to use these different queues ? 

Daniela

**Comment by fstagni** [comment]
The platform is determined at runtime by each pilot : https://github.com/DIRACGrid/Pilot/blob/master/Pilot/pilotCommands.py#L864

Historically, DIRAC does not trust this specific bit of info when it is published by the sites, because we've seen in the past that:
- the information on BDII for certain sites was, and I guess still is somewhere, garbage (admins not knowing what to put...)
- the WNs behind the queue are not always homogeneous, despite what we are supposed to see
- platform is meaningful for your user's applications [this is a point to expand, no time to write a book now]
- for HPCs and other types of opportunistic resources it makes little sense

So, bottom line, you might approach the thing 2 ways:
- do nothing
- understand what really is needed. Are there compatibility issues?

**Reply by marianne013** [reply]
We have users asking for a specific major release (7/8/9). (We do allow the "Any" option for users who think their containerization is solid.) So we need to match these jobs somehow. At the moment the match is on CE level (because that's what we always had) and I was just hoping we could attach this property to the queue. 
@sfayer: Any opinion ?

**Reply by sfayer** [reply]
No particular opinions from me: The pilot side of the matching for this all already works perfectly (users set a Platform tag and get the environment they expect based on what was detected at runtime)... We just want to use the job submission side by setting the platform at the queue level rather than just the CE (similar to how Tags work now).

Regards,
Simon

**Reply by fstagni** [reply]
The WorkloadManagement/Matcher is presented with a dict like:

```
{
CPUTime:               1783296
DIRACVersion:          9.0.0a4
GridCE:                ceprod01.grid.hep.ph.ic.ac.uk
MaxRAM:                23936
NumberOfProcessors:    1
OwnerGroup:            dirac_prod,dteam_admin,dteam_pilot,dteam_user
PilotBenchmark:        25.8
PilotInfoReportedFlag: False
PilotReference:        htcondorce://ceprod01.grid.hep.ph.ic.ac.uk/1521262.0
Platform:              Linux_x86_64_glibc-2.17
ReleaseVersion:        9.0.0a4
Site:                  LCG.UKI-LT2-IC-HEP.uk
Tag:                   Token:gridpp
VirtualOrganization:   dteam
}
```
but that `Platform` that you see there is filled in at RunTime but the Pilot command I linked above.

If your users would specify in their JDLs that the Platform they need is "Linux_x86_64_glibc-2.17" then the matching would work as you expect (and no need to add anything in CS).

**Reply by sfayer** [reply]
Yes, that bit all works perfectly... (We actually have a GridPPDIRAC function that re-writes the Platform with a more friendly alias and that's all fine, users specify the Platform and jobs only run on the Platforms they expect).

The problem is on the submission side with the SiteDirector and CheckPlatform=True (because with small experiments we can't send lots of pilots that might not match). At the moment we have a structure like this in the CS:
```
SiteA
 - CE1
   - Architecture: x86_64
   - OS: Linux_CentOS_7
   - Queues
     - ce-queue-CentOS7
     - ce-queue-Alma8
```

The parameters related to the Platform are at the CE level, but we'd like to have them at the Queue level as the different queues have different operating systems. Is that possible?

Regards,
Simon

**Reply by fstagni** [reply]
You can have "Platform" string at the level of the queue instead of Architecture+OS at the level of the CE: https://github.com/DIRACGrid/DIRAC/blob/34ad06697e5305f813d758cc34467d4908029f64/src/DIRAC/WorkloadManagementSystem/Utilities/QueueUtilities.py#L121

**Comment by fstagni** [comment]
How are users asking for a specific OS? The target OS is not a concept in DIRAC, what we have as concept is "Platform", which is what's returned by the `dirac-platform` command. But that does not include the OS, it's only a string like "Linux_x86_64_glibc-2.34". 

A way of doing what you want ~now is to use the [tag mechanism](https://dirac.readthedocs.io/en/latest/AdministratorGuide/Systems/WorkloadManagement/tagsAndJobs.html#the-generic-tags-mechanism-for-jobs-matching) but that means that:
- on one side you (administrator) manually add specific tags to specific queues.
- on the other, the users while preparing their jobs that add the "Tag"(s) that they want

**Comment by marianne013** [comment, marked as answer]
Our users specify Platform = "EL8"; or Platform = "AnyPlatform" or the equivalent in the python API;  We set in OSCompatibility EL9 = x86_64_EL9 etc, but to make a long story short, we now set "Platform" in the Queues, so RAL-LCG2's EL7 queue has Platform = EL7, while their EL8 queue on the same CE has Platform = EL8 and this seems to result in the desired behaviour. I note that even on the certification server the CEs seem to have an OS specified, is this is not used, maybe it should be dropped ?
